### DISCUSSION TITLE
Problem with Cylindrical FFD

### DISCUSSION THREAD

**Discussion by LuukvdMark** [su2code/SU2#2065] [Q&A]
Hi All,

I'm currently trying to implement a cylindrical FFD box, however, I ran into some problems with the definition of the FFD Box. To explain my problem a bit better I made use of a simplified test case of a simple tube.

My problem is as follows. When I define the cylindrical FFD box with FFD_Degree = 1 for the J_degree everything is fine see included figure (FFDbox_degree_j=1.png). However, I would like to increase the FFD_Degree along this side so I have more control over the deformation on this side. This however causes the FFD box to go towards one side of the circle, see the second included figure (FFDbox_degree_j=2.png). As I was trying to figure out what was happening behind the screens I found this piece of code in CFreeFormDefBox.cpp:

` 
void CFreeFormDefBox::SetCart2Cyl_ControlPoints(CConfig *config) {

  unsigned short iDegree, jDegree, kDegree;
  su2double CartCoord[3];
  su2double X_0, Y_0, Z_0, Xbar, Ybar, Zbar;

  X_0 = config->GetFFD_Axis(0); Y_0 = config->GetFFD_Axis(1);  Z_0 = config->GetFFD_Axis(2);

  for (kDegree = 0; kDegree <= nDegree; kDegree++) {
    for (jDegree = 0; jDegree <= mDegree; jDegree++) {
      for (iDegree = 0; iDegree <= lDegree; iDegree++) {

        CartCoord[0] = Coord_Control_Points[iDegree][jDegree][kDegree][0];
        CartCoord[1] = Coord_Control_Points[iDegree][jDegree][kDegree][1];
        CartCoord[2] = Coord_Control_Points[iDegree][jDegree][kDegree][2];

        Xbar =  CartCoord[0] - X_0; Ybar =  CartCoord[1] - Y_0; Zbar =  CartCoord[2] - Z_0;

        Coord_Control_Points[iDegree][jDegree][kDegree][0] = sqrt(Ybar*Ybar + Zbar*Zbar);
        Coord_Control_Points[iDegree][jDegree][kDegree][1] = atan2 ( Zbar, Ybar);
        if (Coord_Control_Points[iDegree][jDegree][kDegree][1] > PI_NUMBER/2.0) Coord_Control_Points[iDegree][jDegree][kDegree][1] -= 2.0*PI_NUMBER;
        Coord_Control_Points[iDegree][jDegree][kDegree][2] = Xbar;

        CartCoord[0] = Coord_Control_Points_Copy[iDegree][jDegree][kDegree][0];
        CartCoord[1] = Coord_Control_Points_Copy[iDegree][jDegree][kDegree][1];
        CartCoord[2] = Coord_Control_Points_Copy[iDegree][jDegree][kDegree][2];

        Xbar =  CartCoord[0] - X_0; Ybar =  CartCoord[1] - Y_0; Zbar =  CartCoord[2] - Z_0;

        Coord_Control_Points_Copy[iDegree][jDegree][kDegree][0] = sqrt(Ybar*Ybar + Zbar*Zbar);
        Coord_Control_Points_Copy[iDegree][jDegree][kDegree][1] = atan2 (Zbar, Ybar);
        if (Coord_Control_Points_Copy[iDegree][jDegree][kDegree][1] > PI_NUMBER/2.0) Coord_Control_Points_Copy[iDegree][jDegree][kDegree][1] -= 2.0*PI_NUMBER;
        Coord_Control_Points_Copy[iDegree][jDegree][kDegree][2] = Xbar;

      }
    }
  }

} 
`  

I'm not familiar with the language, however, I could deduce that this is where the coordinate point of the cartesian FFD box are translated to cylindrical coordinates. Moreover, I thought maybe the code just had some troubles with the 1 extra point along the J-side, as the r and theta for the cylindrical coordinates would be the same due to symmetry. So I tried to define the FFd box with a larger degree along the J-side, however, the same behaviour showed (see FFDbox_degree_j=2.png)

My question: Am I doing something wrong in the definition of my FFD box? If yes, how could I fix this? If not, is there a way to work around this problem?
![FFDbox_degree_j=1](https://github.com/su2code/SU2/assets/61149282/527b4487-17db-451d-a90e-d2854560f994)
![FFDbox_degree_j=2](https://github.com/su2code/SU2/assets/61149282/0b796e83-eb50-4ac5-a520-2721fbeede22)
![FFDbox_degree_j=3](https://github.com/su2code/SU2/assets/61149282/b1bfac79-0a02-4263-8e15-16cbef32c156)
[config_FFD.txt](https://github.com/su2code/SU2/files/11892864/config_FFD.txt)

**Comment by pcarruscag** [comment]
You have an example here https://github.com/su2code/SU2/blob/develop/TestCases/deformation/cylindrical_ffd/def_cylindrical.cfg with explanation of how to setup the "corners".

**Reply by LuukvdMark** [reply]
Hey pcarruscag,

First of all, thanks for the reply!

I'm aware of the example, and I followed it with my current problem. However, I can't fully check what is done as the mesh for this example is not available. Moreover, I think this example might have the same problem as it also has only 1 degree for the i-axis. When I was playing around with my problem I found that the definition FFD-axis in combination with the mesh and the FFD-box definition can affect whether the i-axis or j-axis shows this problem. I however couldn't get it fixed with different configurations. Any idea what might cause this problem?

**Reply by pcarruscag** [reply]
The mesh is in the test cases repo https://github.com/su2code/TestCases/tree/develop/deformation/cylindrical_ffd
Are you trying to have more points in the radial direction?

**Reply by pcarruscag** [reply]
This is what I get for that case with 3rd-degree FFD in the radial direction
![image](https://github.com/su2code/SU2/assets/38071223/03e24c10-3aa7-4b04-969c-4bad6bd3d0f0)

**Reply by LuukvdMark** [reply]
Thanks for the quick reply again and all the help. I used the mesh of the example file and found what I was doing wrong, I get similar results now.
